### PR TITLE
fix #487: srwpy now uses main SRW repo

### LINKED ISSUE #487: 20230808 srwpy now uses main SRW repo
We can consider using pip install srwpy now, it links to the main SRW repo:

> Sirepo building procedures may need to be changed to just do pip install --pre srwpy. Also, we had to restructure the package, so that it better fits the Python's site-package structure, and now the imports need to be changed from import uti_math to from srwpy import uti_math and so on.

### PR DESCRIPTION
**Pull request by rorour**

### CODE REVIEW

File: installers/rpm-code/codes/srw.sh [hunk 1]
```diff
@@ -33,43 +33,51 @@ srw_main() {
...
-    patch --quiet env/work/srw_python/srwlib.py <<'EOF'
-diff --git a/env/work/srw_python/srwlib.py b/env/work/srw_python/srwlib.py
-index 9e052df..0ab17fb 100644
---- a/env/work/srw_python/srwlib.py
-+++ b/env/work/srw_python/srwlib.py
-@@ -10267,6 +10267,8 @@ def srwl_wfr_emit_prop_multi_e(_e_beam, _mag, _mesh, _sr_meth, _sr_rel_prec, _n_
-                         #srwl_uti_save_intens_ascii(resStokes2.to_deg_coh(), meshRes2, fpdc2, _n_stokes = 1, _arLabels = resLabelsToSaveDC, _arUnits = resUnitsToSaveDC, _mutual = doMutual, _cmplx = 0) #OC12072019 # Deg. of Coh. Cut vs Y
-                         #srwl_uti_save_intens_ascii(resStokes2.to_deg_coh(), meshRes2, fpdc2, _n_stokes = 1, _arLabels = resLabelsToSaveDC, _arUnits = resUnitsToSaveDC, _mutual = 2, _cmplx = 0) #OC16072019 # Deg. of Coh. Cut vs Y
-                         srwl_uti_save_intens(resStokes2.to_deg_coh(), meshRes2, fpdc2, _n_stokes = 1, _arLabels = resLabelsToSaveDC, _arUnits = resUnitsToSaveDC, _mutual = 2, _cmplx = 0, _form = _file_form) #OC17072021 # Deg. of Coh. Cut vs Y
-+                elif _char == 61 and i + 1 < nRecv:
```
Comment (robnagler) [inline installers/rpm-code/codes/srw.sh:55]: This patch is important so you need to look at the code in srwlib.py in the new release to see if we need to patch it.

File: installers/rpm-code/codes/srw.sh [hunk 2]
```diff
@@ -33,43 +33,51 @@ srw_main() {
...
+        uti_mag
+        uti_math
+        uti_math_eigen
+        uti_parse
+        uti_plot
+        uti_plot_com
+        uti_plot_matplotlib
+    )
+
+    cd /home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/
```
Comment (robnagler) [inline installers/rpm-code/codes/srw.sh:64]: We never hardcode paths. There are variables and functions to support this. This one is `$(codes_python_lib_dir)`.

You also don't want to cd to directories unless necessary. Just use declare `d=$(codes_python_lib_dir)` and prefix all paths with this variable.

File: installers/rpm-code/codes/srw.sh [hunk 3]
```diff
@@ -33,43 +33,51 @@ srw_main() {
...
+import srwpy_import_warning
 
-srw_python_patch_uti_math_eigen() {
-    patch --quiet env/work/srw_python/uti_math_eigen.py <<'EOF'
-@@ -29,7 +29,7 @@
- #from sklearn.decomposition import PCA as PCA #OC04082021 (moved to try-except)
- try:
-     from scipy.linalg import eigh as largest_eigh
--    from scipy.sparse.linalg.eigen.arpack import eigsh as largest_eigsh
-+    from scipy.sparse.linalg import eigsh as largest_eigsh
```
Comment (robnagler) [inline installers/rpm-code/codes/srw.sh:70]: I think this patch may be important too.

File: installers/rpm-code/codes/srw.sh [hunk 4]
```diff
@@ -33,43 +33,51 @@ srw_main() {
...
+displayed = False
+
+def check(calling_module):
+    global displayed
+    if not displayed:
+        displayed = True
+        raise DeprecationWarning(f"This method of calling {calling_module} is deprecated. Please change to import from srwpy (ex: import {calling_module} -> from srwpy import {calling_module})")
 EOF
-}
+    for i in "${old_modules[@]}"
```
Comment (robnagler) [inline installers/rpm-code/codes/srw.sh:74]: we use semicolon style and declare i:

```sh
declare i
for i in "${old_modules[@]}"; do
```

File: installers/rpm-code/codes/srw.sh [hunk 5]
```diff
@@ -33,43 +33,51 @@ srw_main() {
...
+    do
+        cat > $i.py <<EOF
```
Comment (robnagler) [inline installers/rpm-code/codes/srw.sh:76]: We probably want to use install here. There are examples in codes. In this case:

```sh
install -m 444 /dev/stdin $d/$i.py <<EOF
```

install ensures permissions and ownership are correct.

File: installers/rpm-code/codes/srw.sh [hunk 6]
```diff
@@ -33,43 +33,51 @@ srw_main() {
...
+
+    cd /home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/
+        cat > srwpy_import_warning.py <<EOF
+displayed = False
+
+def check(calling_module):
+    global displayed
+    if not displayed:
+        displayed = True
+        raise DeprecationWarning(f"This method of calling {calling_module} is deprecated. Please change to import from srwpy (ex: import {calling_module} -> from srwpy import {calling_module})")
```
Comment (robnagler) [inline installers/rpm-code/codes/srw.sh:72]: does the `raise` terminate the caller?

`This method of importing`.

File: installers/rpm-code/codes/srw.sh [hunk 7]
```diff
@@ -33,43 +33,51 @@ srw_main() {
 
 srw_python_install() {
     install_pip_install primme
-    cd SRW/cpp/py
```
Comment (robnagler) [inline installers/rpm-code/codes/srw.sh:36]: srw_main does not need to do the make. That's done by srw_python_install.

Also, please verify that the pip installed *.so correctly links to sfftw.

**Review by robnagler — CHANGES_REQUESTED** [review]

File: installers/rpm-code/codes/srw.sh (hunk 7 above)
Comment (rorour) [reply installers/rpm-code/codes/srw.sh:36]: > srw_main does not need to do the make. That's done by srw_python_install.

Please explain this more, do you mean I should remove line 31?

> Also, please verify that the pip installed *.so correctly links to sfftw.

How can I verify this?

Comment (robnagler) [reply installers/rpm-code/codes/srw.sh:36]: > Please explain this more, do you mean I should remove line 31?

Lines 25-31. pip install is doing the make in its setup.py.

> How can I verify this?

This is with the current rpm:

```sh
[py3;@v site-packages]$ ldd srwlpy.so
ldd: warning: you do not have execution permission for `./srwlpy.so'
        linux-vdso.so.1 (0x00007ffd4a76c000)
        libfftw3f.so.3 => /lib64/libfftw3f.so.3 (0x00007f2ed9600000)
        libfftw3.so.3 => /lib64/libfftw3.so.3 (0x00007f2ed940c000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f2ed9000000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f2ed932e000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f2ed930e000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f2ed8c00000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f2ed99fc000)
```

libfftw3f is the single precision library. I'm not sure what's doing the conversion from sfftw to fftw3f. I think we need to look to see if the makefile is getting generated correctly with the old version; does sfttw actually get injected with the perl command or is there a bug here that means setup.py does the right thing?

Comment (rorour) [reply installers/rpm-code/codes/srw.sh:36]: Let's meet to discuss this and the patches.

File: installers/rpm-code/codes/srw.sh (hunk 1 above)
Comment (rorour) [reply installers/rpm-code/codes/srw.sh]: The existing patch works on the new file. 994d56ec4eb7008c310ccbcf15d10cae03f86f38

File: installers/rpm-code/codes/srw.sh (hunk 2 above)
Comment (rorour) [reply installers/rpm-code/codes/srw.sh]: 994d56ec4eb7008c310ccbcf15d10cae03f86f38

File: installers/rpm-code/codes/srw.sh (hunk 3 above)
Comment (rorour) [reply installers/rpm-code/codes/srw.sh:70]: patch not needed 994d56ec4eb7008c310ccbcf15d10cae03f86f38

File: installers/rpm-code/codes/srw.sh (hunk 4 above)
Comment (rorour) [reply installers/rpm-code/codes/srw.sh]: 994d56ec4eb7008c310ccbcf15d10cae03f86f38

File: installers/rpm-code/codes/srw.sh (hunk 5 above)
Comment (rorour) [reply installers/rpm-code/codes/srw.sh]: 994d56ec4eb7008c310ccbcf15d10cae03f86f38

File: installers/rpm-code/codes/srw.sh (hunk 6 above)
Comment (rorour) [reply installers/rpm-code/codes/srw.sh]: Yes, I changed to a print. https://github.com/radiasoft/download/commit/994d56ec4eb7008c310ccbcf15d10cae03f86f38

File: installers/rpm-code/codes/srw.sh (hunk 7 above)
Comment (rorour) [reply installers/rpm-code/codes/srw.sh:36]: https://github.com/radiasoft/download/commit/994d56ec4eb7008c310ccbcf15d10cae03f86f38

File: installers/rpm-code/codes/srw.sh [hunk 8]
```diff
@@ -60,16 +49,47 @@ index 9e052df..0ab17fb 100644
...
+
+    declare d=$(codes_python_lib_dir)
+    install -m 444 /dev/stdin $d/srwpy_import_warning.py <<EOF
+displayed = False
+
+def check(calling_module):
+    global displayed
+    if not displayed:
+        displayed = True
+        print(f"This method of importing {calling_module} is deprecated. Please change to import from srwpy (ex: import {calling_module} -> from srwpy import {calling_module})")
```
Comment (robnagler) [inline installers/rpm-code/codes/srw.sh:84]: add `file=sys.stderr` arg to print so stdout isn't corrupted by warning.

**Review by robnagler — CHANGES_REQUESTED** [review]

File: installers/rpm-code/codes/srw.sh (hunk 8 above)
Comment (rorour) [reply installers/rpm-code/codes/srw.sh]: 8c0f8d7fbcbe9043ee23049f39fca0719946fe1a

**Pull request merged**